### PR TITLE
Signup route

### PR DESCRIPTION
**Pull request by MohammadKoolivand**
stablishing a signup route contain (helper file, controller and validator )

### CODE REVIEW

File: src/controllers/app/userController.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+const User = require("../../models/user");
+const bcrypt = require("bcrypt");
+const utilities = require("../../helpers/utilities");
+const moment = require("moment");
```
Comment (farshid0594) [inline src/controllers/app/userController.js:4]: fix imports modules:
first import absoulte modules 
second import relative modules.

File: src/controllers/app/userController.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    // get req body
+    var { username, password, mobile } = req.body;
+    // check user exists
+    User.findOne({ username: username }, (err, findedUser) => {
+      if (err) {
+        return res.status(500).json({ message: "server error" });
+      }
+      if (findedUser) {
+        return res.status(400).json({ message: "User exists" });
+      }
```
Comment (farshid0594) [inline src/controllers/app/userController.js:17]: if user exist but not active. What happen?

File: src/controllers/app/userController.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+      // hash password
+      bcrypt.hash(password, bcrypt.genSaltSync(10), (err, hash) => {
+        if (err) {
+          return res.status(500).json({ message: err });
```
Comment (farshid0594) [inline src/controllers/app/userController.js:22]: dont return err.

File: src/controllers/app/userController.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        // generate Code
+        const generatedCode = utilities.generateCode();
+
+        // save new user
+        var newUser = new User({
+          username: username,
+          password: hash,
+          mobile: mobile,
+          code: {
+            codeNo: generatedCode,
```
Comment (farshid0594) [inline src/controllers/app/userController.js:34]: change codeNo to text.!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: src/controllers/app/userController.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            expired: moment().add(2, "minute").format("YYYYMMDDThhmmss"),
+          },
+        });
+        newUser.save((err, savedNewUser) => {
+          if (err) {
+            return res
+              .status(500)
+              .json({ message: "Server Error", err: err });
```
Comment (farshid0594) [inline src/controllers/app/userController.js:42]: don`t return err

File: src/controllers/app/userController.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+          }
+          return res.status(200).json({
+            message: "successfully saved",
+            id: savedNewUser._id,
+            code: generatedCode,
```
Comment (farshid0594) [inline src/controllers/app/userController.js:47]: why return the code?

File: src/controllers/app/userController.js (hunk 1 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js:4]: done

File: src/helpers/utils.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+exports.generateCode = function () {
+  return Math.floor(Math.random() * 9000) + 1000;
+};
```
Comment (farshid0594) [inline src/helpers/utils.js:3]: chnage to :
exports.generateCode = function (len=4) {
    var code = ""
    for (var i = 0; i < len; i++) {
        code = code + Math.floor(Math.random() * 10)
    }
    return code
}

File: src/models/User.js [hunk 8]
```diff
@@ -1,21 +1,20 @@
...
-    {
-        timestamps: true,
-    }
+  {
+    username: { type: String, required: true, unique: true, index: true },
+    password: { type: String, required: true },
+    mobile: { type: String, required: true, unique: true },
+    name: { type: String, maxlength: 30 },
+    tokens: [{ text: String, expired: String }],
+    code: { text: Number, expired: String },
```
Comment (farshid0594) [inline src/models/User.js]: change type text to string

File: src/controllers/app/userController.js (hunk 6 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js:47]: deleted

File: src/routes/app/index.js [hunk 9]
```diff
@@ -4,9 +4,19 @@ const express = require("express");
...
+//import validator
+const LoginValidator = require("../../validations/loginValidate");
+
+
+//import controller
+const UserController = require("../../controllers/app/userController");
+
+
 //routes
 AppRouter.get("/", (req, res) => {
```
Comment (farshid0594) [inline src/routes/app/index.js:16]: remove this route

File: src/validations/loginValidate.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
+exports.ValidateSignUp = (req, res, next) => {
+  var errors = [];
```
Comment (farshid0594) [inline src/validations/loginValidate.js:2]: chnage var to const

File: src/validations/loginValidate.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+exports.ValidateSignUp = (req, res, next) => {
+  var errors = [];
+  var { username, password } = req.body;
```
Comment (farshid0594) [inline src/validations/loginValidate.js:3]: chnage var to const

File: src/controllers/app/userController.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
+const User = require("../../models/user");
+const bcrypt = require("bcrypt");
+const utilities = require("../../helpers/utilities");
+const moment = require("moment");
+
+exports.SignUp = function (req, res) {
+  try {
+    // get req body
+    var { username, password, mobile } = req.body;
```
Comment (farshid0594) [inline src/controllers/app/userController.js:9]: chnage var to const

File: src/validations/loginValidate.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+exports.ValidateSignUp = (req, res, next) => {
+  var errors = [];
+  var { username, password } = req.body;
+  if (password.length < 6) {
```
Comment (farshid0594) [inline src/validations/loginValidate.js:4]: If there is no password, no error will be returned.
same for username

File: src/helpers/utils.js (hunk 7 above)
Comment (MohammadKoolivand) [reply src/helpers/utils.js:3]: done

File: src/controllers/app/userController.js (hunk 12 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js:9]: done

File: src/validations/loginValidate.js (hunk 11 above)
Comment (MohammadKoolivand) [reply src/validations/loginValidate.js:3]: done

File: src/validations/loginValidate.js (hunk 10 above)
Comment (MohammadKoolivand) [reply src/validations/loginValidate.js:2]: done

File: src/controllers/app/userController.js (hunk 3 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js:22]: done

File: src/controllers/app/userController.js (hunk 4 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js:34]: done done done

File: src/controllers/app/userController.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+const bcrypt = require("bcrypt");
+const utilities = require("../../helpers/utilities");
+const moment = require("moment");
+
+exports.SignUp = function (req, res) {
+  try {
+    // get req body
+    var { username, password, mobile } = req.body;
+    // check user exists
+    User.findOne({ username: username }, (err, findedUser) => {
```
Comment (farshid0594) [inline src/controllers/app/userController.js:11]: If the mobile is duplicate, What happened?

File: src/routes/app/index.js (hunk 9 above)
Comment (MohammadKoolivand) [reply src/routes/app/index.js:16]: done

File: src/models/User.js (hunk 8 above)
Comment (MohammadKoolivand) [reply src/models/User.js]: done

File: src/controllers/app/userController.js (hunk 5 above)
Comment (MohammadKoolivand) [reply src/controllers/app/userController.js:42]: done

**Review by farshid0594 — APPROVED** [review]

**Pull request closed without merging**